File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/signus/CreateMyDidTest.java
package org.hyperledger.indy.sdk.signus;

import org.hyperledger.indy.sdk.IndyIntegrationTestWithSingleWallet;
import org.hyperledger.indy.sdk.InvalidStructureException;
import org.hyperledger.indy.sdk.signus.SignusResults.CreateAndStoreMyDidResult;

import static org.hamcrest.CoreMatchers.isA;
import static org.junit.Assert.assertEquals;

import org.bitcoinj.core.Base58;
import org.junit.Test;

import java.util.concurrent.ExecutionException;

public class CreateMyDidTest extends IndyIntegrationTestWithSingleWallet {

	@Test
	public void testCreateMyDidWorksForEmptyJson() throws Exception {
		CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(wallet, "{}").get();

		assertEquals(16, Base58.decode(result.getDid()).length);
		assertEquals(32, Base58.decode(result.getVerkey()).length);
	}

	@Test
	public void testCreateMyDidWorksForSeed() throws Exception {
		CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(this.wallet, MY1_IDENTITY_JSON).get();

		assertEquals(DID_FOR_MY1_SEED, result.getDid());
		assertEquals(VERKEY_FOR_MY1_SEED, result.getVerkey());
	}

	@Test
	public void testCreateMyDidWorksAsCid() throws Exception {
		String  didJson = new SignusJSONParameters.CreateAndStoreMyDidJSONParameter(null, MY1_SEED, null, true).toJson();

		CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(this.wallet, didJson).get();

		assertEquals(VERKEY_FOR_MY1_SEED, result.getDid());
		assertEquals(VERKEY_FOR_MY1_SEED, result.getVerkey());
	}

	@Test
	public void testCreateMyDidWorksForPassedDid() throws Exception {

		String didJson = new SignusJSONParameters.CreateAndStoreMyDidJSONParameter(DID1, null, null, false).toJson();

		CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(this.wallet, didJson).get();

		assertEquals(DID1, result.getDid());
	}

	@Test
	public void testCreateMyDidWorksForCorrectCryptoType() throws Exception {
		String didJson = new SignusJSONParameters.CreateAndStoreMyDidJSONParameter(null, MY1_SEED, CRYPTO_TYPE, null).toJson();

		CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(this.wallet, didJson).get();

		assertEquals(DID_FOR_MY1_SEED, result.getDid());
		assertEquals(VERKEY_FOR_MY1_SEED, result.getVerkey());
	}

	@Test
	public void testCreateMyDidWorksForInvalidSeed() throws Exception {
		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(InvalidStructureException.class));

		String didJson = new SignusJSONParameters.CreateAndStoreMyDidJSONParameter(null, "aaaaaaaaaaa", null, null).toJson();

		Signus.createAndStoreMyDid(this.wallet, didJson).get();
	}

	@Test
	public void testCreateMyDidWorksForInvalidCryptoType() throws Exception {
		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(UnknownCryptoException.class));

		String didJson = new SignusJSONParameters.CreateAndStoreMyDidJSONParameter(null, MY1_SEED, "crypto_type", null).toJson();

		Signus.createAndStoreMyDid(this.wallet, didJson).get();
	}

	@Test
	public void testCreateMyDidWorksForAllParams() throws Exception {
		String didJson = new SignusJSONParameters.CreateAndStoreMyDidJSONParameter(DID1, MY1_SEED, CRYPTO_TYPE, true).toJson();

		CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(this.wallet, didJson).get();

		assertEquals(DID1, result.getDid());
		assertEquals(VERKEY_FOR_MY1_SEED, result.getVerkey());
	}
}

File: wrappers/dotnet/indy-sdk-dotnet-test/DidTests/CreateMyDidTest.cs
using Hyperledger.Indy.DidApi;
using Hyperledger.Indy.Test.Util.Base58Check;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Threading.Tasks;

namespace Hyperledger.Indy.Test.SignusTests
{
    [TestClass]
    public class CreateMyDidTest : IndyIntegrationTestWithSingleWallet
    {        
        [TestMethod]
        public async Task TestCreateMyDidWorksForEmptyJson()
        {
            var result = await Did.CreateAndStoreMyDidAsync(wallet, "{}");
            Assert.IsNotNull(result);

            Assert.AreEqual(16, Base58CheckEncoding.DecodePlain(result.Did).Length);
            Assert.AreEqual(32, Base58CheckEncoding.DecodePlain(result.VerKey).Length);
        }

        [TestMethod]
        public async Task TestCreateMyDidWorksForSeed()
        {
            var result = await Did.CreateAndStoreMyDidAsync(wallet, MY1_IDENTITY_JSON);
            Assert.IsNotNull(result);

            Assert.AreEqual(DID_MY1, result.Did);
            Assert.AreEqual(VERKEY_MY1, result.VerKey);
        }

        [TestMethod]
        public async Task TestCreateMyDidWorksAsCid()
        {
            var json = string.Format("{{\"seed\":\"{0}\",\"cid\":true}}", MY1_SEED);

            var result = await Did.CreateAndStoreMyDidAsync(wallet, json);
            Assert.IsNotNull(result);

            Assert.AreEqual(VERKEY_MY1, result.Did);
            Assert.AreEqual(VERKEY_MY1, result.VerKey);
        }

        [TestMethod]
        public async Task TestCreateMyDidWorksForPassedDid()
        {
            var json = string.Format("{{\"did\":\"{0}\",\"cid\":false}}", DID1);

            var result = await Did.CreateAndStoreMyDidAsync(wallet, json);
            Assert.IsNotNull(result);

            Assert.AreEqual(DID1, result.Did);
        }

        [TestMethod]
        public async Task TestCreateMyDidWorksForCorrectCryptoType()
        {
            var json = string.Format("{{\"seed\":\"{0}\",\"crypto_type\":\"{1}\"}}", MY1_SEED, CRYPTO_TYPE);

            var result = await Did.CreateAndStoreMyDidAsync(wallet, json);


            Assert.AreEqual(DID_MY1, result.Did);
            Assert.AreEqual(VERKEY_MY1, result.VerKey); 
        }

        [TestMethod]
        public async Task TestCreateMyDidWorksForInvalidSeed()
        {
            var json = "{\"seed\":\"aaaaaaaaaaa\"}";

            var ex = await Assert.ThrowsExceptionAsync<InvalidStructureException>(() =>
                Did.CreateAndStoreMyDidAsync(wallet, json)
            );
        }

        [TestMethod]
        public async Task TestCreateMyDidWorksForInvalidCryptoType()
        {
            var json = string.Format("{{\"seed\":\"{0}\",\"crypto_type\":\"crypto_type\"}}", MY1_SEED);

            var ex = await Assert.ThrowsExceptionAsync<UnknownCryptoException>(() =>
                Did.CreateAndStoreMyDidAsync(wallet, json)
            );
        }

        [TestMethod]
        public async Task TestCreateMyDidWorksForAllParams()
        {
            var json = string.Format("{{\"did\":\"{0}\",\"seed\":\"{1}\",\"crypto_type\":\"{2}\",\"cid\":true}}", DID1, MY1_SEED, CRYPTO_TYPE);

            var result = await Did.CreateAndStoreMyDidAsync(wallet, json);
            Assert.IsNotNull(result);

            Assert.AreEqual(DID1, result.Did);
            Assert.AreEqual(VERKEY_MY1, result.VerKey);
        }

    }
}
